File: members-android/app/src/main/java/ch/fwvraura/members/data/model/EventMaterials.kt
package ch.fwvraura.members.data.model

import com.google.gson.annotations.SerializedName
import java.util.Locale

/**
 * Modelle fuer das editierbare Event-Modul "Rezepte & Material" (Organisator und
 * Vorstand). Gegenstueck zur read-only Anzeige in Recipe.kt.
 *
 * Zahlenwerte koennen vom Backend als String ODER Number kommen (PG-numeric) —
 * daher werden alle numerischen Anzeigefelder als String? gehalten und ueber
 * [EventMaterialNumbers.toDouble] tolerant geparst.
 */

/** Ein mit dem Event verknuepftes Rezept (GET events/{id}/recipes, ausfuehrlich). */
data class LinkedRecipe(
    val id: Int? = null,
    @SerializedName("recipe_id") val recipeId: Int? = null,
    @SerializedName("link_id") val linkId: Int? = null,
    val name: String? = null,
    /** Portionen fuer diesen Anlass. */
    val servings: String? = null,
    @SerializedName("category_name") val categoryName: String? = null,
    /** Aktuell aus dem Lager herstellbare Portionen. */
    @SerializedName("available_portions") val availablePortions: String? = null
) {
    /** Rezept-ID fuer PUT/DELETE: bevorzugt recipe_id, sonst id. */
    val effectiveRecipeId: Int? get() = recipeId ?: id
}

/** Ein auswaehlbares Rezept (GET events/{id}/available-recipes). */
data class AvailableRecipe(
    val id: Int,
    val name: String? = null,
    @SerializedName("category_name") val categoryName: String? = null,
    @SerializedName("available_portions") val availablePortions: String? = null
)

/** Ein auswaehlbares Material (GET events/{id}/available-items). */
data class AvailableItem(
    val id: Int,
    val name: String? = null,
    val unit: String? = null,
    val quantity: String? = null,
    @SerializedName("category_name") val categoryName: String? = null
)

/** Body von POST events/{id}/recipes — Rezept verknuepfen (Upsert von Portionen). */
data class LinkRecipeRequest(
    @SerializedName("recipe_id") val recipeId: Int,
    val servings: Int
)

/** Body von PUT events/{id}/recipes/{recipeId} — Portionen aendern. */
data class UpdateServingsRequest(val servings: Int)

/** Body von POST events/{id}/manual-items — manuelles Material hinzufuegen. */
data class AddManualItemRequest(
    @SerializedName("item_id") val itemId: Int,
    val quantity: Double
)

/** Tolerante Zahl- und Formathilfen fuer das Rezepte-&-Material-Modul. */
object EventMaterialNumbers {
    /** Parst eine Zahl, die als String ODER Number geliefert wurde. */
    fun toDouble(value: String?): Double = value?.trim()?.toDoubleOrNull() ?: 0.0

    /** Menge ohne unnoetige Nachkommastellen (2.0 → "2", 1.5 → "1.5"/"1.50"). */
    fun format(value: Double): String =
        if (value % 1.0 == 0.0) value.toLong().toString()
        else String.format(Locale.US, "%.2f", value)

    /** "{Menge} {Einheit}" bzw. nur die Menge, wenn keine Einheit vorhanden. */
    fun formatQty(value: Double, unit: String?): String {
        val n = format(value)
        return if (unit.isNullOrBlank()) n else "$n $unit"
    }

    /** Betrag immer mit zwei Nachkommastellen (fuer CHF-Summen). */
    fun formatCost(value: Double): String = String.format(Locale.US, "%.2f", value)
}

File: vorstand-desktop/src/lib/types/event-materials.ts
// Typen fuer das Event-Modul "Rezepte & Material". Zahlenwerte koennen vom
// Backend als String ODER Number geliefert werden (PG-numeric) — daher immer
// `Number(x)` vor Rechnen/`toFixed`. Deshalb hier `number | string`.

type Numeric = number | string;

/** Ein mit dem Event verknuepftes Rezept (GET /events/{id}/recipes). */
export interface EventRecipe {
  /** Rezept-ID (r.id). Fuer PUT/DELETE genutzt. */
  id: number;
  /** Manche Antworten liefern zusaetzlich recipe_id — dann bevorzugen. */
  recipe_id?: number;
  /** ID der Verknuepfung (event_recipes.id) — nur informativ. */
  link_id?: number;
  name: string;
  /** Portionen fuer diesen Anlass. */
  servings: Numeric;
  category_name?: string | null;
  /** Aktuell aus dem Lager herstellbare Portionen. */
  available_portions?: Numeric | null;
}

/** Ein auswaehlbares Rezept (GET /events/{id}/available-recipes). */
export interface AvailableRecipe {
  id: number;
  name: string;
  category_name?: string | null;
  available_portions?: Numeric | null;
}

/** Ein auswaehlbares Material (GET /events/{id}/available-items). */
export interface AvailableItem {
  id: number;
  name: string;
  unit?: string | null;
  quantity?: Numeric | null;
  category_name?: string | null;
}

/** Eine Position der Einkaufsliste (GET /events/{id}/shopping-list). */
export interface ShoppingListItem {
  item_id: number;
  item_name: string;
  unit?: string | null;
  /** Gesamtbedarf = recipe_needed + manual_needed. */
  needed: Numeric;
  /** Bedarf aus verknuepften Rezepten. */
  recipe_needed: Numeric;
  /** Bedarf aus manuell erfassten Positionen. */
  manual_needed: Numeric;
  /** Aktueller Lagerbestand. */
  in_stock: Numeric;
  /** Fehlmenge (needed - in_stock, min. 0) — das, was eingekauft werden muss. */
  to_buy: Numeric;
  purchase_price?: Numeric | null;
  supplier?: string | null;
  estimated_cost?: Numeric | null;
  purchased?: boolean;
}

/** Antwort der Einkaufsliste inkl. Summen. */
export interface ShoppingList {
  event_slug?: string;
  items: ShoppingListItem[];
  total_items: number;
  total_to_buy: number;
  estimated_total_cost: Numeric;
  estimated_open_cost: Numeric;
}
